### ISSUE TITLE
Has anyone tried to use Edge TPU with Minigo?

### ISSUE THREAD

**Issue by iopq** [tensorflow/minigo#824]
So, can it actually pay games with current models?

**Comment by tommadams** [comment]
Yes, it absolutely can. Instructions on how to train a compatible model are here: https://coral.withgoogle.com/docs/edgetpu/models-intro/

For best results however, we need to train a smaller model that can fit entirely on the Edge TPU. Our latest model is too large to fit entirely on the Edge TPU on-board memory, so it needs to stream weights over, which slows performance considerably.

I'm going to retrain a smaller model using our v17 games as examples this week. For best results, we need to retrain a model anyway because the it needs to be quantized (though we don't need to do any selfplay so this shouldn't take too long).

**Comment by tommadams** [comment]
To give some slightly more concrete numbers:

Full size model (19 trunk layers, 256 filter width):

- GPU: inference for one position takes 14ms.
- Edge TPU: inference for one position takes 38ms.

With 15 trunk layers, 160 filter width:

- GPU: one position takes 5.1ms.
- Edge TPU: one position takes 6.7ms.

With 19 trunk layers, 128 filter width:

- GPU: one position takes 3.8ms.
- Edge TPU: one position takes 4.0ms.


For the GPU runs, I used the Minigo C++ engine running on an NVIDIA Quadro M2000 with virtual losses set to 8 (because performance is very bad with vlosses == 1). The model was 32bit float.

For the Edge TPU runs, I used the Minigo Python engine and virtual losses disabled (there's no performance advantage to using it and it adversely affects tree search quality). The model was 8bit.

You can see when the model is small enough to fit Edge TPU's on-board memory, performance is comparable to a > $400 GPU (though this isn't exactly fair to the GPU since it's running a 32bit model).

These numbers came from running a model bootstrapped with random noise so I don't have a measure of how quantization affects the model's strength yet.

**Comment by iopq** [comment]
Cool, I have two questions:

1. How does it fare vs. the GPU in matches?
2. Is there any difference between the board model and the usb model, do you know?

**Comment by tommadams** [comment]
1. I don't know yet. In order to run a model on the Edge TPU, it has to be quantized first. This means retraining the model. I tried retraining a model yesterday from a sample of the latest 20% of the selfplay games from our v17 model, but the value head didn't seem to learn correctly. Our first guess as to why this might be is that the model learns some important things about how to "count" during the early games that are a lot simpler. I started another training run this afternoon using a different sampling strategy and I'll have the results tomorrow morning.

2. I can't comment on performance between the two models just yet because I've only tried a prototype dev board. I'm hopefully getting my hands on a consumer dev board soon and will be able to test then. To my knowledge, the Edge TPU chips are basically the same on both. The main difference between them is that the Edge TPU is connected to the CPU via PCIe on the dev board, rather than USB. I don't know what the difference in bandwidth is, it might be somewhere in the [data sheet](https://coral.withgoogle.com/docs/dev-board/datasheet/).

**Comment by iopq** [comment]
We're all watching closely for updates!

**Comment by tommadams** [comment]
Sorry, I've been super busy with other things the last couple of weeks. I did have time to retrain a quantized model from our v17 dataset, which you can get from GCS at `gs://minigo-pub/edgetpu-19x19/minigo-v17-2019-04-29-edgetpu.tflite` or by running the following command:

```
curl http://storage.googleapis.com/minigo-pub/edgetpu-19x19/minigo-v17-2019-04-29-edgetpu.tflite -o saved_models/minigo-v17-2019-04-29-edgetpu.tflite
```

The model has 19 trunk layers and a filter width of 128. Currently, we've only integrated edge tpu support into Minigo's Python engine.

I haven't had time to do a thorough comparison of the smaller quantized model vs the full v17 model.

**Comment by tommadams** [comment]
A guide to running Minigo on EdgeTPU was published a couple of weeks ago: https://coral.withgoogle.com/projects/minigo

**Comment by alreadydone** [comment]
"The AlphaGo Zero algorithm from DeepMind shocked the AI community when it became the first AI to beat a professional go player, using a relatively simple approach."

The shock is that AGZ uses a relatively simple approach "without human knowledge", not that it beats a pro for the first time. It indeed crushed AGL which beat a top pro, though.

Nice to know Google Seattle is researching plasma fusion.

**Comment by LBDonovan** [comment]
Hi @tommadams 

I really appreciate the work you've put in to get Minigo running on the edge tpu, it's a remarkable achievement!

I'm working on a project which will interface Minigo running on a Coral dev board to a physical 9x9 go board, reading the game state using a camera and outputting moves via an LED matrix embedded within the board.

It seems that the only publicly available trained and compiled Minigo model compatible with the edge tpu is for a 19x19 board. As I understand the size of the board can not be modified at runtime and so it would be necessary to retrain a quantised model from scratch for a 9x9 board. Is this understanding correct?

If so could you briefly talk me through the steps you would take to train an edge-tpu-compatible 9x9 model on a GCP instance, so that I know I'm heading in the right direction?

Many thanks!

**Comment by amj** [comment]
@LBDonovan Your understanding is correct!

We have information about how to train a model specifically for an edge-tpu here: https://coral.ai/projects/minigo/#training-your-own-go-models

One difficulty is that minigo hasn't created/published 9x9 training data, so this data would need to be created via selfplay, or sourced some other way.  There may be good selfplay datasets that already exist, such as at [CGOS](http://www.yss-aya.com/cgos/) [(more info)](http://www.yss-aya.com/ayaself/ayaself.html).  We have some early bad data available in [the GCS bucket
here](https://console.cloud.google.com/storage/browser/minigo-pub/v3-9x9), but it may not load still.

Hope this helps!
